client/app/services/query-snippet.js

```diff
@@ -13,17 +17,16 @@ function QuerySnippetService($resource) {
       content: this.snippet,
       tabTrigger: this.trigger,
     };
-  };
-
-  return resource;
+  }
 }
 
-export default function init(ngModule) {
-  ngModule.factory("QuerySnippet", QuerySnippetService);
+const getQuerySnippet = querySnippet => new QuerySnippet(querySnippet);
 
-  ngModule.run($injector => {
-    QuerySnippet = $injector.get("QuerySnippet");
-  });
-}
+const QuerySnippetService = {
+  query: () => axios.get("api/query_snippets").then(data => map(data, getQuerySnippet)),
+  create: data => axios.post("api/query_snippets", data).then(getQuerySnippet),
+  save: data => axios.post(`api/query_snippets/${data.id}`, data).then(getQuerySnippet),
```

Comment:
Interesting solution :+1: But you should keep in mind that some API calls do not return a full updated instance (for performance reasons), but only updated fields. I believe, only query and dashboard API do so, but anyway. Also, in some places we started using a (sort of) pattern: we implement a method that updates some instance; it accepts instance and a list of updates, makes a request to backend, then creates a copy of instance and picks from server's response only fields passed in arguments (or few others, like `version` for queries and dashboard). I think we should not try to mimic Angular resource here and try to use this pattern where possible.

Reply:
I think the simpler ones don't need much more than this.

My thoughts for this part was to try to keep every response returning a new valid object and minimizing the changes in code. The update pattern is interesting, but also depends on each resource... Wdyt of simplifying it here and postpone this update to when we define a standard for an ideal api resource layer?

It's not "exactly" an Angular resource mimic :sweat_smile:, the class in this case was more to show how it could be done, only the API requests change the object, so in this case it's not an issue :). Angular resources also had the api methods in its prototype (`$save`, `$delete`, etc), those are quickly changeable, I'm more concerned about the ones that have lots of extra stuff in the prototype (I'll share comments when I reach them :slightly_smiling_face:)